kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/FilterInvokers.java

```diff
@@ -44,17 +44,26 @@ static List<FilterAndInvoker> from(String filterName, Filter filter) {
     }
 
     private static List<FilterAndInvoker> invokersForFilter(String filterName, Filter filter) {
-        boolean isResponseFilter = filter instanceof ResponseFilter;
-        boolean isRequestFilter = filter instanceof RequestFilter;
-        boolean isAnySpecificFilterInterface = SpecificFilterArrayInvoker.implementsAnySpecificFilterInterface(filter);
-        validateFilter(filterName, isResponseFilter, isRequestFilter, isAnySpecificFilterInterface);
-        if (isResponseFilter && isRequestFilter) {
+        validateFilter(filterName, filter);
+        boolean isAnyRequestFilter = filter instanceof RequestFilter;
+        boolean isAnyResponseFilter = filter instanceof ResponseFilter;
+        boolean isSpecificRequestFilter = SpecificFilterArrayInvoker.implementsAnySpecificRequestFilterInterface(filter);
+        boolean isSpecificResponseFilter = SpecificFilterArrayInvoker.implementsAnySpecificResponseFilterInterface(filter);
+        if (isAnyRequestFilter && isAnyResponseFilter) {
             return singleFilterAndInvoker(filterName, filter, new RequestResponseInvoker((RequestFilter) filter, (ResponseFilter) filter));
         }
-        else if (isRequestFilter) {
+        else if (isAnyRequestFilter && isSpecificResponseFilter) {
+            return List.of(new FilterAndInvoker(filterName, filter, new RequestFilterInvoker((RequestFilter) filter)),
+                    new FilterAndInvoker(filterName, filter, arrayInvoker(filter)));
+        }
+        else if (isSpecificRequestFilter && isAnyResponseFilter) {
+            return List.of(new FilterAndInvoker(filterName, filter, arrayInvoker(filter)),
+                    new FilterAndInvoker(filterName, filter, new ResponseFilterInvoker((ResponseFilter) filter)));
+        }
```

Comment:
why have two branches here? if one FilterInvoker cares about responses and one cares about requests the order shouldn't matter.

Reply:
I don't understand what you mean here @robobario

Reply:
can't you do:

```
else if ((isAnyRequestFilter && isSpecificResponseFilter) || (isSpecificRequestFilter && isAnyResponseFilter)) {
            return List.of(new FilterAndInvoker(filterName, filter, new RequestFilterInvoker((RequestFilter) filter)),
                    new FilterAndInvoker(filterName, filter, arrayInvoker(filter)));
        }
```

Reply:
here's a refactor PR https://github.com/tombentley/kproxy/pull/23

I think it catches the intent better by deduplicating where we create the specific invoker, and deduplicates the interface testing across validation/instantiation without having to pass a bunch of boolean args.